File: w3_p1/0.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int main() {
	char c[101];
	int i,j,m,n,a,b;
	char num[13][7] ={"2","3","4","5","6","7","8","9","10","Jack","Queen","King","Ace"};
	char shape[4][10]={"Clubs","Diamonds","Hearts","Spades"};
	int arr[100][55];
	int arr1[55];
	int arr2[55];

	while(scanf("%d", &m)==1)
	while(m--){
		scanf("%d",&n);
		for(i = 1;i<=n;i++){
			for(j=1;j<=52;j++){
				scanf("%d", &arr[i][j]);
			}
		}
		for(i=1;i<=52;i++){
			arr1[i] = i;
		}
		getchar();
		while(gets(c)&&c[0]){
			sscanf(c,"%d",&a);
			for(i = 1; i<=52;i++){
				arr2[i] = arr1[arr[a][i]];
			}
			for(i=1;i<=52;i++){
				arr1[i] = arr2[i];
			}
		}
		for(i =1;i<=52;i++){
			b=arr1[i]-1;
			printf("%s of %s\n", num[b%13],shape[b/13]);
		}
		if(m){
			puts("");
		}
	}
	return 0;
}

File: w3_p1/0.py
# -*- coding: utf-8 -*-
# UTF-8 encoding when using korean
import sys

x = int(input())
t = [2,3,4,5,6,7,8,9,10,'Jack','Queen','King','Ace']
shape = ['Clubs','Diamonds','Hearts','Spades']
n = input()
for v in range(x):

	n = int(input())
	l = []
	for _ in range(n):
		l.append(list(map(int,input().split())))
	
	p = [i for i in range(1,53)]
	
	k = sys.stdin.readline().rstrip('\n')
	while k != '':
		k = int(k)
		f = [False] * 52
		for j in range(52):
			f[j] = p[l[k-1][j]-1]
		p = f
		k = sys.stdin.readline().rstrip('\n')
	
	for i in p:
		if (i-1) // 13 == 4:
			a = shape[3]
		else:
			a = shape[(i-1) // 13]
		b = t[(i-1) % 13]
		print('{0} of {1}'.format(b,a))
	
	if v != x-1:
		print()